Improved type of callable-array

robchett opened the pull request:
BC: Removes TCallableArray & TCallableList in favour of TCallableKeyedArray
Closes #7778

Also improves the type of callable-array so that it is interpreted as a `list{class-string|object, string}` instead of `non-empty-array<array-key, mixed>`

robchett commented:
@orklah what would be preferred here

  -'callable-array{0: class-string, 1: string}'
  +'callable-array{class-string, string}'

As this is now a list of size 2 I'm not sure which better defines the shape.

orklah commented:
Semantically, the second syntax implies that the order is locked while the first doesn't (meaning `[1=>'method', 0=>stdClass::class]` would be valid, but it shouldn't). So the second syntax is best

orklah commented:
I see TCallableArray being removed but I don't see TCallableList being removed as well, did I miss something?

robchett reviewed (commented):
TCallableList was removed upstream - remove BC note about it

robchett commented:
I rebased this branch last week as it was a long way behind. Looks like TCallableList was removed elsewhere upstream. Tracked that down to cca276768ef2f3e91e3763afdc32d16cc2b4da03, removed the note about that in the BC note

orklah commented on src/Psalm/Internal/Type/SimpleAssertionReconciler.php:
```diff
@@ -2651,10 +2650,6 @@ private static function reconcileCallable(
             ) {
                 $callable_types[] = $type;
                 $redundant = false;
-            } elseif ($type instanceof TArray) {
```
Shouldn't this create a TCallableKeyedArray now that TCallableArray doesn't exists?

robchett replied:
Yes, that would make more sense and I don't remember why I made that change to begin with. It doesn't appear to effect any of the tests either way which

orklah commented on src/Psalm/Internal/Type/TypeParser.php:
```diff
@@ -1411,7 +1411,6 @@ private static function getTypeFromKeyedArrayTree(
                         !is_numeric($property_key)
                         || ($had_optional && !$property_maybe_undefined)
                         || $type === 'array'
-                        || $type === 'callable-array'
```
why remove this? TCallableKeyedArray still have a `callable-array` key, no?

robchett replied:
I have no good reason why it was removed. Now back and tests are unaffected,

orklah commented:
Thanks!
